spec/compiler/crystal/tools/doc/project_info_spec.cr

```diff
@@ -0,0 +1,118 @@
+require "../../../../spec_helper"
+require "../../../../support/tempfile"
+
+private alias ProjectInfo = Crystal::Doc::ProjectInfo
+
+describe Crystal::Doc::ProjectInfo do
+  it ".new_with_defaults" do
+    with_tempfile("docs-defaults") do |tempdir|
+      Dir.mkdir tempdir
+      Dir.cd(tempdir) do
+        # Empty dir
+        ProjectInfo.new_with_defaults(nil, nil) { |name, version| "missing:#{name}:#{version}" }.should eq "missing::"
+        ProjectInfo.new_with_defaults("foo", "1.0") { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "1.0")
+
+        # shard.yml
+        File.write("shard.yml", "name: foo\nversion: 1.0")
+        ProjectInfo.new_with_defaults(nil, nil) { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "1.0")
+        ProjectInfo.new_with_defaults("bar", "2.0") { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("bar", "2.0")
+        ProjectInfo.new_with_defaults(nil, "2.0") { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "2.0")
+
+        # git tagged version
+        `git init`
+        `git add shard.yml`
+        `git commit -m 'Initial commit' --no-gpg-sign`
+        `git tag v3.0`
+        ProjectInfo.new_with_defaults(nil, nil) { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "3.0")
+        ProjectInfo.new_with_defaults("bar", "2.0") { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("bar", "2.0")
+
+        # git dirty dir
+        File.write("foo.txt", "bar")
+        ProjectInfo.new_with_defaults(nil, nil) { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "1.0")
```

Comment:
This case needs some refinement. It feels wrong to simply fall back to the shard.yml version if the git repo is dirty. This is likely incorrect and a source of confusion. To make a dirty repo more explicit, I'd suggest to append `-dev` to the version (or some other indicator if there are better suggestions). In that specific case, I would still use the tag from HEAD (and append `-dev`) because git tags should have priority over shard.yml property.
If we're not in a git repo, using shard.yml version is fine.